packages/jsii-pacmak/bin/jsii-pacmak.ts

```diff
@@ -78,192 +87,153 @@ import { VERSION_DESC } from '../lib/version';
 
-  async function generateTarget(assembly: reflect.Assembly, packageDir: string, targetName: string, targetOutputDir: string, tarball: string) {
+  logging.info(`Packaged. ${timers.display()}`);
+
+  async function buildTargetsForLanguage(targetLanguage: string, modules: JsiiModule[], perLanguageDirectory: boolean) {
     // ``argv.target`` is guaranteed valid by ``yargs`` through the ``choices`` directive.
-    const targetConstructor = targetConstructors[targetName];
-    if (!targetConstructor) {
-      throw new Error(`Unsupported target: "${targetName}"`);
+    const builder = ALL_BUILDERS[targetLanguage as TargetName];
+    if (!builder) {
+      throw new Error(`Unsupported target: '${targetLanguage}'`);
     }
 
-    const target = new targetConstructor({
-      targetName,
-      packageDir,
-      assembly,
-      fingerprint: argv.fingerprint,
+    await builder.buildModules(modules, {
+      clean: argv.clean,
+      codeOnly: argv['code-only'],
       force: argv.force,
-      arguments: argv
+      fingerprint: argv.fingerprint,
+      arguments: argv,
+      languageSubdirectory: perLanguageDirectory,
     });
-
-    const codeDir = argv.codeOnly ? targetOutputDir : await fs.mkdtemp(path.join(os.tmpdir(), 'jsii-pacmak-code'));
-
-    logging.debug(`Generating ${targetName} code into ${codeDir}`);
-
-    await target.generateCode(codeDir, tarball);
-
-    if (argv.codeOnly) { return; }
-
-    logging.debug(`Building into ${targetOutputDir}`);
-    await target.build(codeDir, targetOutputDir);
-
-    if (argv.clean) {
-      await fs.remove(codeDir);
-    } else {
-      logging.info(`Generated code for ${targetName} retained at: ${codeDir}`);
-    }
   }
-
 })().catch(err => {
   process.stderr.write(`${err.stack}\n`);
   process.exit(1);
 });
 
-async function npmPack(packageDir: string, tmpdir: string): Promise<string> {
-  logging.debug(`Running "npm pack ${packageDir}" in ${tmpdir}`);
-  const args = ['pack', packageDir];
-  if (logging.level >= logging.LEVEL_VERBOSE) {
-    args.push('--loglevel=verbose');
-  }
-  const out = await shell('npm', args, { cwd: tmpdir });
-  // Take only the last line of npm pack which should contain the
-  // tarball name. otherwise, there can be a lot of extra noise there
-  // from scripts that emit to STDOUT.
-  const lines = out.trim().split(os.EOL);
-  return path.resolve(tmpdir, lines[lines.length - 1].trim());
+/**
+ * A set of packages (targets) translated into the same language
+ */
+interface TargetSet {
+  targetType: string;
+  modules: JsiiModule[];
 }
 
-async function updateNpmIgnore(packageDir: string, excludeOutdir: string | undefined) {
-  const npmIgnorePath = path.join(packageDir, '.npmignore');
-  let lines = new Array<string>();
-  let modified = false;
-  if (await fs.pathExists(npmIgnorePath)) {
-    lines = (await fs.readFile(npmIgnorePath)).toString().split('\n');
+function sliceTargets(modules: JsiiModule[], requestedTargets: string[] | undefined, force: boolean) {
```

Comment:
I think this should make sure all available modules support all requested targets (or only return targets supported by all). If we brought in additional modules for dependencies, and some dependencies don't support a target, then the build is doomed to fail anyway...

Reply:
Could be, although to do that properly we would have to record the dependencies (not all packages are necessarily dependencies of each other).

So unrelated packages A and B could look like:

```
A: java+python
B: python
```

And I might want to build all targets. Though unlikely, I did not see a reason to prohibit that kind of build. Since the goal of this is speed, independent builds would require analyzing A twice, which would take away some of the purpose.